test/Microsoft.AspNetCore.Server.KestrelTests/PipelineExtensionTests.cs

```diff
@@ -0,0 +1,166 @@
+        [InlineData(_ulongMaxValueLength - 1)]
+        public void WritesNumericAcrossSpanBoundaries(int gapSize)
+        {
+            var writer = _pipe.Writer.Alloc(100);
+            // almost fill up the first block
+            var spacer = new Span<byte>(new byte[writer.Buffer.Length - gapSize]);
+            writer.Write(spacer);
+
+            var bufferLength = writer.Buffer.Length;
+            writer.WriteNumeric(ulong.MaxValue);
+            Assert.NotEqual(bufferLength, writer.Buffer.Length);
+
+            writer.FlushAsync().GetAwaiter().GetResult();
+
+            var reader = _pipe.Reader.ReadAsync().GetAwaiter().GetResult();
+            var numAsString = ulong.MaxValue.ToString();
+            var written = reader.Buffer.Slice(spacer.Length, numAsString.Length);
+            Assert.False(written.IsSingleSpan, "The buffer should cross spans");
+            AssertExtensions.Equal(Encoding.ASCII.GetBytes(numAsString), written.ToArray());
+        }
+
+        [Theory]
+        [InlineData("\0abcxyz", new byte[] { 0, 97, 98, 99, 120, 121, 122 })]
+        [InlineData("!#$%i", new byte[] { 33, 35, 36, 37, 105 })]
+        [InlineData("!#$%", new byte[] { 33, 35, 36, 37 })]
+        [InlineData("!#$", new byte[] { 33, 35, 36 })]
+        [InlineData("!#", new byte[] { 33, 35 })]
+        [InlineData("!", new byte[] { 33 })]
+        // null or empty
+        [InlineData("", new byte[0])]
+        [InlineData(null, new byte[0])]
+        public async Task EncodesAsAscii(string input, byte[] expected)
+        {
+            var writer = _pipe.Writer.Alloc();
+            writer.WriteAsciiNoValidation(input);
+            await writer.FlushAsync();
+            var reader = await _pipe.Reader.ReadAsync();
+
+            if (expected.Length > 0)
+            {
+                AssertExtensions.Equal(
+                    expected,
+                    reader.Buffer.ToArray());
+            }
+            else
+            {
+                Assert.Equal(0, reader.Buffer.Length);
+            }
+        }
+
+        [Theory]
+        // non-ascii characters stored in 32 bits
+        [InlineData("𤭢𐐝")]
+        // non-ascii characters stored in 16 bits
+        [InlineData("ñ٢⛄⛵")]
+        public async Task WriteAsciiNoValidationWritesOnlyOneBytePerChar(string input)
+        {
+            // WriteAscii doesn't validate if characters are in the ASCII range
+            // but it shouldn't produce more than one byte per character
+            var writer = _pipe.Writer.Alloc();
+            writer.WriteAsciiNoValidation(input);
+            await writer.FlushAsync();
+            var reader = await _pipe.Reader.ReadAsync();
+
+            Assert.Equal(input.Length, reader.Buffer.Length);
+        }
+
+        [Fact]
+        public async Task WriteAsciiNoValidation()
+        {
+            const byte maxAscii = 0x7f;
+            var writer = _pipe.Writer.Alloc();
+            for (var i = 0; i < maxAscii; i++)
```

Comment:
Seems unnecessary to make it a theory. The test only has one assert. So if it goes wrong, the assert error would be 
```
Assert.Equal() Failure
Expected: (x)
Actual: (y)
```